File: ProgressHUD/ProgressHUD/ProgressHUD.m
//
// Copyright (c) 2018 Related Code - http://relatedcode.com
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#import "ProgressHUD.h"

@implementation ProgressHUD

@synthesize window, viewBackground, toolbarHUD, spinner, imageView, labelStatus, timer;
@synthesize fontStatus, colorStatus, colorSpinner, colorHUD, colorBackground, imageSuccess, imageError;

//-------------------------------------------------------------------------------------------------------------------------------------------------
+ (ProgressHUD *)shared
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	static dispatch_once_t once;
	static ProgressHUD *progressHUD;
	//---------------------------------------------------------------------------------------------------------------------------------------------
	dispatch_once(&once, ^{ progressHUD = [[ProgressHUD alloc] init]; });
	//---------------------------------------------------------------------------------------------------------------------------------------------
	return progressHUD;
}

#pragma mark - Display methods

//-------------------------------------------------------------------------------------------------------------------------------------------------
+ (void)dismiss
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	dispatch_async(dispatch_get_main_queue(), ^{
		[[self shared] hudHide];
	});
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
+ (void)show
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	dispatch_async(dispatch_get_main_queue(), ^{
		[[self shared] hudCreate:nil image:nil spin:YES hide:NO interaction:YES];
	});
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
+ (void)show:(NSString *)status
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	dispatch_async(dispatch_get_main_queue(), ^{
		[[self shared] hudCreate:status image:nil spin:YES hide:NO interaction:YES];
	});
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
+ (void)show:(NSString *)status Interaction:(BOOL)interaction
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	dispatch_async(dispatch_get_main_queue(), ^{
		[[self shared] hudCreate:status image:nil spin:YES hide:NO interaction:interaction];
	});
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
+ (void)showSuccess
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	dispatch_async(dispatch_get_main_queue(), ^{
		[[self shared] hudCreate:nil image:[self shared].imageSuccess spin:NO hide:YES interaction:YES];
	});
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
+ (void)showSuccess:(NSString *)status
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	dispatch_async(dispatch_get_main_queue(), ^{
		[[self shared] hudCreate:status image:[self shared].imageSuccess spin:NO hide:YES interaction:YES];
	});
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
+ (void)showSuccess:(NSString *)status Interaction:(BOOL)interaction
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	dispatch_async(dispatch_get_main_queue(), ^{
		[[self shared] hudCreate:status image:[self shared].imageSuccess spin:NO hide:YES interaction:interaction];
	});
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
+ (void)showError
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	dispatch_async(dispatch_get_main_queue(), ^{
		[[self shared] hudCreate:nil image:[self shared].imageError spin:NO hide:YES interaction:YES];
	});
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
+ (void)showError:(NSString *)status
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	dispatch_async(dispatch_get_main_queue(), ^{
		[[self shared] hudCreate:status image:[self shared].imageError spin:NO hide:YES interaction:YES];
	});
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
+ (void)showError:(NSString *)status Interaction:(BOOL)interaction
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	dispatch_async(dispatch_get_main_queue(), ^{
		[[self shared] hudCreate:status image:[self shared].imageError spin:NO hide:YES interaction:interaction];
	});
}

#pragma mark - Property methods

//-------------------------------------------------------------------------------------------------------------------------------------------------
+ (void)fontStatus:(UIFont *)font			{ [self shared].fontStatus		= font;	 }
+ (void)colorStatus:(UIColor *)color		{ [self shared].colorStatus		= color; }
+ (void)colorSpinner:(UIColor *)color		{ [self shared].colorSpinner	= color; }
+ (void)colorHUD:(UIColor *)color			{ [self shared].colorHUD		= color; }
+ (void)colorBackground:(UIColor *)color	{ [self shared].colorBackground	= color; }
+ (void)imageSuccess:(UIImage *)image		{ [self shared].imageSuccess	= image; }
+ (void)imageError:(UIImage *)image			{ [self shared].imageError		= image; }
//-------------------------------------------------------------------------------------------------------------------------------------------------

#pragma mark -

//-------------------------------------------------------------------------------------------------------------------------------------------------
- (id)init
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	self = [super initWithFrame:[[UIScreen mainScreen] bounds]];
	//---------------------------------------------------------------------------------------------------------------------------------------------
	UIColor *labelColor = [UIColor blackColor];
	if (@available(iOS 13, *)) { labelColor = [UIColor labelColor]; }
	//---------------------------------------------------------------------------------------------------------------------------------------------
	fontStatus			= [UIFont boldSystemFontOfSize:16];
	colorStatus			= labelColor;
	colorSpinner		= labelColor;
	colorHUD			= [UIColor grayColor];
	colorBackground		= [UIColor colorWithRed:0.0 green:0.0 blue:0.0 alpha:0.2];;
	NSBundle *bundle	= [NSBundle bundleForClass:[self class]];
	imageSuccess		= [UIImage imageNamed:@"ProgressHUD.bundle/progresshud-success" inBundle:bundle compatibleWithTraitCollection:nil];
	imageError			= [UIImage imageNamed:@"ProgressHUD.bundle/progresshud-error" inBundle:bundle compatibleWithTraitCollection:nil];
	//---------------------------------------------------------------------------------------------------------------------------------------------
	id<UIApplicationDelegate> delegate = [[UIApplication sharedApplication] delegate];
	//---------------------------------------------------------------------------------------------------------------------------------------------
	if ([delegate respondsToSelector:@selector(window)])
		window = [delegate performSelector:@selector(window)];
	else window = [[[UIApplication sharedApplication] windows] firstObject];
	//---------------------------------------------------------------------------------------------------------------------------------------------
	self.alpha = 0;
	//---------------------------------------------------------------------------------------------------------------------------------------------
	return self;
}

#pragma mark -

//-------------------------------------------------------------------------------------------------------------------------------------------------
- (void)hudCreate:(NSString *)status image:(UIImage *)image spin:(BOOL)spin hide:(BOOL)hide interaction:(BOOL)interaction
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	if (toolbarHUD == nil)
	{
		toolbarHUD = [[UIToolbar alloc] initWithFrame:CGRectZero];
		toolbarHUD.translucent = YES;
		toolbarHUD.backgroundColor = colorHUD;
		toolbarHUD.layer.cornerRadius = 10;
		toolbarHUD.layer.masksToBounds = YES;
		[self registerNotifications];
	}
	//---------------------------------------------------------------------------------------------------------------------------------------------
	if (toolbarHUD.superview == nil)
	{
		if (interaction == NO)
		{
			viewBackground = [[UIView alloc] initWithFrame:window.frame];
			viewBackground.backgroundColor = colorBackground;
			[window addSubview:viewBackground];
			[viewBackground addSubview:toolbarHUD];
		}
		else [window addSubview:toolbarHUD];
	}
	//---------------------------------------------------------------------------------------------------------------------------------------------
	if (spinner == nil)
	{
		UIActivityIndicatorViewStyle style = UIActivityIndicatorViewStyleWhiteLarge;
		if (@available(iOS 13, *)) { style = UIActivityIndicatorViewStyleLarge; }
		spinner = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:style];
		spinner.color = colorSpinner;
		spinner.hidesWhenStopped = YES;
	}
	if (spinner.superview == nil) [toolbarHUD addSubview:spinner];
	//---------------------------------------------------------------------------------------------------------------------------------------------
	if (imageView == nil)
	{
		imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, 28, 28)];
	}
	if (imageView.superview == nil) [toolbarHUD addSubview:imageView];
	//---------------------------------------------------------------------------------------------------------------------------------------------
	if (labelStatus == nil)
	{
		labelStatus = [[UILabel alloc] initWithFrame:CGRectZero];
		labelStatus.font = fontStatus;
		labelStatus.textColor = colorStatus;
		labelStatus.backgroundColor = [UIColor clearColor];
		labelStatus.textAlignment = NSTextAlignmentCenter;
		labelStatus.baselineAdjustment = UIBaselineAdjustmentAlignCenters;
		labelStatus.numberOfLines = 0;
	}
	if (labelStatus.superview == nil) [toolbarHUD addSubview:labelStatus];
	//---------------------------------------------------------------------------------------------------------------------------------------------

	//---------------------------------------------------------------------------------------------------------------------------------------------
	labelStatus.text = status;
	labelStatus.hidden = (status == nil) ? YES : NO;
	//---------------------------------------------------------------------------------------------------------------------------------------------
	imageView.image = image;
	imageView.hidden = (image == nil) ? YES : NO;
	//---------------------------------------------------------------------------------------------------------------------------------------------
	if (spin) [spinner startAnimating]; else [spinner stopAnimating];
	//---------------------------------------------------------------------------------------------------------------------------------------------

	//---------------------------------------------------------------------------------------------------------------------------------------------
	[self hudSize];
	[self hudPosition:nil];
	[self hudShow];
	//---------------------------------------------------------------------------------------------------------------------------------------------
	if (hide) [self timedHide];
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
- (void)registerNotifications
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(hudPosition:) name:UIDeviceOrientationDidChangeNotification object:nil];
	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(hudPosition:) name:UIKeyboardWillHideNotification object:nil];
	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(hudPosition:) name:UIKeyboardWillShowNotification object:nil];
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
- (void)hudDestroy
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	[[NSNotificationCenter defaultCenter] removeObserver:self];
	//---------------------------------------------------------------------------------------------------------------------------------------------
	[labelStatus removeFromSuperview];		labelStatus = nil;
	[imageView removeFromSuperview];		imageView = nil;
	[spinner removeFromSuperview];			spinner = nil;
	[toolbarHUD removeFromSuperview];		toolbarHUD = nil;
	[viewBackground removeFromSuperview];	viewBackground = nil;
	//---------------------------------------------------------------------------------------------------------------------------------------------
	if (timer != nil) { [timer invalidate]; timer = nil; }
}

#pragma mark -

//-------------------------------------------------------------------------------------------------------------------------------------------------
- (void)hudSize
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	CGRect rectLabel = CGRectZero;
	CGFloat widthHUD = 100, heightHUD = 100;
	//---------------------------------------------------------------------------------------------------------------------------------------------
	if (labelStatus.text != nil)
	{
		NSDictionary *attributes = @{NSFontAttributeName:labelStatus.font};
		NSInteger options = NSStringDrawingUsesFontLeading | NSStringDrawingTruncatesLastVisibleLine | NSStringDrawingUsesLineFragmentOrigin;
		rectLabel = [labelStatus.text boundingRectWithSize:CGSizeMake(200, 300) options:options attributes:attributes context:NULL];

		widthHUD = rectLabel.size.width + 50;
		heightHUD = rectLabel.size.height + 75;

		if (widthHUD < 100) widthHUD = 100;
		if (heightHUD < 100) heightHUD = 100;

		rectLabel.origin.x = (widthHUD - rectLabel.size.width) / 2;
		rectLabel.origin.y = (heightHUD - rectLabel.size.height) / 2 + 25;
	}
	//---------------------------------------------------------------------------------------------------------------------------------------------
	toolbarHUD.bounds = CGRectMake(0, 0, widthHUD, heightHUD);
	//---------------------------------------------------------------------------------------------------------------------------------------------
	CGFloat imageX = widthHUD/2;
	CGFloat imageY = (labelStatus.text == nil) ? heightHUD/2 : 36;
	imageView.center = spinner.center = CGPointMake(imageX, imageY);
	//---------------------------------------------------------------------------------------------------------------------------------------------
	labelStatus.frame = rectLabel;
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
- (void)hudPosition:(NSNotification *)notification
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	CGFloat heightKeyboard = 0;
	NSTimeInterval duration = 0;
	//---------------------------------------------------------------------------------------------------------------------------------------------
	if (notification != nil)
	{
		NSDictionary *info = [notification userInfo];
		CGRect keyboard = [[info valueForKey:UIKeyboardFrameEndUserInfoKey] CGRectValue];
		duration = [[info valueForKey:UIKeyboardAnimationDurationUserInfoKey] doubleValue];
		if ((notification.name == UIKeyboardWillShowNotification) || (notification.name == UIKeyboardDidShowNotification))
		{
			heightKeyboard = keyboard.size.height;
		}
	}
	else heightKeyboard = [self keyboardHeight];
	//---------------------------------------------------------------------------------------------------------------------------------------------
	CGRect screen = [UIScreen mainScreen].bounds;
	CGPoint center = CGPointMake(screen.size.width/2, (screen.size.height-heightKeyboard)/2);
	//---------------------------------------------------------------------------------------------------------------------------------------------
	[UIView animateWithDuration:duration delay:0 options:UIViewAnimationOptionAllowUserInteraction animations:^{
		self.toolbarHUD.center = CGPointMake(center.x, center.y);
	} completion:nil];
	//---------------------------------------------------------------------------------------------------------------------------------------------
	if (viewBackground != nil) viewBackground.frame = window.frame;
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
- (CGFloat)keyboardHeight
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	for (UIWindow *testWindow in [[UIApplication sharedApplication] windows])
	{
		if ([[testWindow class] isEqual:[UIWindow class]] == NO)
		{
			for (UIView *possibleKeyboard in [testWindow subviews])
			{
				if ([[possibleKeyboard description] hasPrefix:@"<UIPeripheralHostView"])
				{
					return possibleKeyboard.bounds.size.height;
				}
				else if ([[possibleKeyboard description] hasPrefix:@"<UIInputSetContainerView"])
				{
					for (UIView *hostKeyboard in [possibleKeyboard subviews])
					{
						if ([[hostKeyboard description] hasPrefix:@"<UIInputSetHost"])
						{
							return hostKeyboard.frame.size.height;
						}
					}
				}
			}
		}
	}
	return 0;
}

#pragma mark -

//-------------------------------------------------------------------------------------------------------------------------------------------------
- (void)hudShow
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	if (timer != nil) [timer invalidate];

	if (self.alpha == 0)
	{
		self.alpha = 1;
		toolbarHUD.alpha = 0;
		toolbarHUD.transform = CGAffineTransformScale(toolbarHUD.transform, 1.4, 1.4);

		UIViewAnimationOptions options = UIViewAnimationOptionAllowUserInteraction | UIViewAnimationCurveEaseOut;
		[UIView animateWithDuration:0.15 delay:0 options:options animations:^{
			self.toolbarHUD.transform = CGAffineTransformScale(self.toolbarHUD.transform, 1/1.4, 1/1.4);
			self.toolbarHUD.alpha = 1;
		} completion:nil];
	}
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
- (void)hudHide
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	if (self.alpha == 1)
	{
		UIViewAnimationOptions options = UIViewAnimationOptionAllowUserInteraction | UIViewAnimationCurveEaseIn;
		[UIView animateWithDuration:0.15 delay:0 options:options animations:^{
			self.toolbarHUD.transform = CGAffineTransformScale(self.toolbarHUD.transform, 0.7, 0.7);
			self.toolbarHUD.alpha = 0;
		}
		completion:^(BOOL finished) {
			[self hudDestroy];
			self.alpha = 0;
		}];
	}
}

//-------------------------------------------------------------------------------------------------------------------------------------------------
- (void)timedHide
//-------------------------------------------------------------------------------------------------------------------------------------------------
{
	NSTimeInterval delay = labelStatus.text.length * 0.04 + 0.5;
	timer = [NSTimer scheduledTimerWithTimeInterval:delay repeats:NO block:^(NSTimer *timer) {
		[self hudHide];
	}];
}

@end

File: ProgressHUD/Sources/ProgressHUD.swift
//
// Copyright (c) 2024 Related Code - https://relatedcode.com
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import UIKit

public class ProgressHUD: UIView {

	var main: UIWindow!

	// Banner properties
	var viewBanner: UIToolbar?
	var colorBanner = UIColor.clear
	var timerBanner: Timer?
	var observerBanner: NSObjectProtocol?

	var textBannerTitle = ""
	var colorBannerTitle = UIColor.label
	var fontBannerTitle = UIFont.boldSystemFont(ofSize: 16)
	var labelBannerTitle: UILabel?

	var textBannerMessage = ""
	var colorBannerMessage = UIColor.secondaryLabel
	var fontBannerMessage = UIFont.systemFont(ofSize: 14)
	var labelBannerMessage: UILabel?

	// HUD properties
	var timerHUD: Timer?

	var mediaSize: CGFloat = 70
	var marginSize: CGFloat = 30

	var viewBackground: UIView?
	var toolbarHUD: UIToolbar?
	var labelStatus: UILabel?

	var viewProgress: ProgressView?
	var viewLiveIcon: UIView?
	var viewStaticImage: UIImageView?
	var viewAnimation: UIView?

	var animationType	= AnimationType.activityIndicator
	var animationSymbol	= "sun.max"

	var colorBackground	= UIColor(red: 0, green: 0, blue: 0, alpha: 0.2)
	var colorHUD		= UIColor.systemGray
	var colorStatus		= UIColor.label
	var colorProgress	= UIColor.lightGray
	var colorAnimation	= UIColor.lightGray

	var fontStatus		= UIFont.boldSystemFont(ofSize: 24)
	var imageSuccess	= UIImage.checkmark.withTintColor(UIColor.systemGreen, renderingMode: .alwaysOriginal)
	var imageError		= UIImage.remove.withTintColor(UIColor.systemRed, renderingMode: .alwaysOriginal)

	var didSetupNotifications	= false
	let keyboardWillShow		= UIResponder.keyboardWillShowNotification
	let keyboardWillHide		= UIResponder.keyboardWillHideNotification
	let keyboardDidShow			= UIResponder.keyboardDidShowNotification
	let keyboardDidHide			= UIResponder.keyboardDidHideNotification
	let orientationDidChange	= UIDevice.orientationDidChangeNotification

	static let shared = ProgressHUD()

	convenience private init() {
		self.init(frame: UIScreen.main.bounds)
		self.alpha = 0
	}

	required internal init?(coder: NSCoder) {
		super.init(coder: coder)
	}

	override private init(frame: CGRect) {
		super.init(frame: frame)
	}
}

// MARK: - Progress
extension ProgressHUD {

	func progress(text: String?, value: CGFloat, interaction: Bool) {

		removeDelayTimer()

		setupWindow()
		setupBackground(interaction)
		setupToolbar()
		setupStatus(text)

		removeLiveIcon()
		removeStaticImage()
		removeAnimationView()
		setupProgressView(value)

		setupSizes(text, false)
		setupNotifications()
		setupPosition()
		displayHUD()
	}
}

// MARK: - Live Icon
extension ProgressHUD {

	func liveIcon(text: String?, icon: LiveIcon, interaction: Bool, delay: TimeInterval?) {

		removeDelayTimer()

		setupWindow()
		setupBackground(interaction)
		setupToolbar()
		setupStatus(text)

		removeStaticImage()
		removeProgressView()
		removeAnimationView()
		setupLiveIcon(icon)
		setupDelayTimer(text, delay)

		setupSizes(text, false)
		setupNotifications()
		setupPosition()
		displayHUD()
	}
}

// MARK: - Static Image
extension ProgressHUD {

	func staticImage(text: String?, image: UIImage?, interaction: Bool, delay: TimeInterval?) {

		removeDelayTimer()

		setupWindow()
		setupBackground(interaction)
		setupToolbar()
		setupStatus(text)

		removeLiveIcon()
		removeProgressView()
		removeAnimationView()
		setupStaticImage(image)
		setupDelayTimer(text, delay)

		setupSizes(text, false)
		setupNotifications()
		setupPosition()
		displayHUD()
	}
}

// MARK: - Animation
extension ProgressHUD {

	func animate(text: String?, interaction: Bool) {

		removeDelayTimer()

		setupWindow()
		setupBackground(interaction)
		setupToolbar()
		setupStatus(text)

		removeLiveIcon()
		removeStaticImage()
		removeProgressView()
		setupAnimationView()

		setupSizes(text, true)
		setupNotifications()
		setupPosition()
		displayHUD()
	}
}

// MARK: - Delay Timer
extension ProgressHUD {

	private func removeDelayTimer() {
		timerHUD?.invalidate()
		timerHUD = nil
	}

	private func setupDelayTimer(_ text: String?, _ delay: TimeInterval?) {
		let count = text?.count ?? 0
		let delay = delay ?? Double(count) * 0.03 + 1.25

		timerHUD = Timer.scheduledTimer(withTimeInterval: delay, repeats: false) { [weak self] _ in
			guard let self = self else { return }
			dismissHUD()
		}
	}
}

// MARK: - Notifications
extension ProgressHUD {

	private func removeNotifications() {
		if (didSetupNotifications) {
			NotificationCenter.default.removeObserver(self)
			didSetupNotifications = false
		}
	}

	private func setupNotifications() {
		if (!didSetupNotifications) {
			NotificationCenter.default.addObserver(self, selector: #selector(setupPosition(_:)), name: keyboardWillShow, object: nil)
			NotificationCenter.default.addObserver(self, selector: #selector(setupPosition(_:)), name: keyboardWillHide, object: nil)
			NotificationCenter.default.addObserver(self, selector: #selector(setupPosition(_:)), name: keyboardDidShow, object: nil)
			NotificationCenter.default.addObserver(self, selector: #selector(setupPosition(_:)), name: keyboardDidHide, object: nil)
			NotificationCenter.default.addObserver(self, selector: #selector(setupPosition(_:)), name: orientationDidChange, object: nil)
			didSetupNotifications = true
		}
	}
}

// MARK: - Window
extension ProgressHUD {

	func setupWindow() {
		if (main == nil) {
			main = UIApplication.shared.windows.first
		}
	}
}

// MARK: - Background
extension ProgressHUD {

	private func removeBackground() {
		viewBackground?.removeFromSuperview()
		viewBackground = nil
	}

	private func setupBackground(_ interaction: Bool) {
		if (viewBackground == nil) {
			viewBackground = UIView(frame: bounds)
			main.addSubview(viewBackground!)
		}

		viewBackground?.backgroundColor = interaction ? .clear : colorBackground
		viewBackground?.isUserInteractionEnabled = !interaction
	}
}

// MARK: - Toolbar
extension ProgressHUD {

	private func removeToolbar() {
		toolbarHUD?.removeFromSuperview()
		toolbarHUD = nil
	}

	private func setupToolbar() {
		if (toolbarHUD == nil) {
			toolbarHUD = UIToolbar(frame: CGRect.zero)
			toolbarHUD?.isTranslucent = true
			toolbarHUD?.clipsToBounds = true
			toolbarHUD?.layer.cornerRadius = 10
			toolbarHUD?.layer.masksToBounds = true
			viewBackground?.addSubview(toolbarHUD!)
		}

		toolbarHUD?.backgroundColor = colorHUD
	}
}

// MARK: - Status Label
extension ProgressHUD {

	private func removeStatus() {
		labelStatus?.removeFromSuperview()
		labelStatus = nil
	}

	private func setupStatus(_ text: String?) {
		if (labelStatus == nil) {
			labelStatus = UILabel()
			labelStatus?.textAlignment = .center
			labelStatus?.baselineAdjustment = .alignCenters
			labelStatus?.numberOfLines = 0
			toolbarHUD?.addSubview(labelStatus!)
		}

		labelStatus?.text = text
		labelStatus?.font = fontStatus
		labelStatus?.textColor = colorStatus
		labelStatus?.isHidden = (text == nil) ? true : false
	}
}

// MARK: - Progress View
extension ProgressHUD {

	private func removeProgressView() {
		viewProgress?.removeFromSuperview()
		viewProgress = nil
	}

	private func setupProgressView(_ value: CGFloat) {
		if (viewProgress == nil) {
			viewProgress = ProgressView(colorProgress)
			viewProgress?.frame = CGRect(x: 0, y: 0, width: mediaSize, height: mediaSize)
		}

		guard let viewProgress = viewProgress else { return }

		if (viewProgress.superview == nil) {
			toolbarHUD?.addSubview(viewProgress)
		}

		viewProgress.setProgress(value)
	}
}

// MARK: - Live Icon
extension ProgressHUD {

	private func removeLiveIcon() {
		viewLiveIcon?.removeFromSuperview()
		viewLiveIcon = nil
	}

	private func setupLiveIcon(_ icon: LiveIcon) {
		if (viewLiveIcon == nil) {
			viewLiveIcon = UIView(frame: CGRect(x: 0, y: 0, width: mediaSize, height: mediaSize))
		}

		guard let viewLiveIcon = viewLiveIcon else { return }

		if (viewLiveIcon.superview == nil) {
			toolbarHUD?.addSubview(viewLiveIcon)
		}

		viewLiveIcon.layer.sublayers?.forEach {
			$0.removeFromSuperlayer()
		}

		if (icon == .succeed)	{ liveIconSucceed(viewLiveIcon)	}
		if (icon == .failed)	{ liveIconFailed(viewLiveIcon)	}
		if (icon == .added)		{ liveIconAdded(viewLiveIcon)	}
	}
}

// MARK: - Static Image
extension ProgressHUD {

	private func removeStaticImage() {
		viewStaticImage?.removeFromSuperview()
		viewStaticImage = nil
	}

	private func setupStaticImage(_ image: UIImage?) {
		if (viewStaticImage == nil) {
			viewStaticImage = UIImageView(frame: CGRect(x: 0, y: 0, width: mediaSize, height: mediaSize))
		}

		guard let viewStaticImage = viewStaticImage else { return }

		if (viewStaticImage.superview == nil) {
			toolbarHUD?.addSubview(viewStaticImage)
		}

		viewStaticImage.image = image
		viewStaticImage.contentMode = .scaleAspectFit
	}
}

// MARK: - Animation View
extension ProgressHUD {

	private func removeAnimationView() {
		viewAnimation?.removeFromSuperview()
		viewAnimation = nil
	}

	private func setupAnimationView() {
		if (viewAnimation == nil) {
			viewAnimation = UIView(frame: CGRect(x: 0, y: 0, width: mediaSize, height: mediaSize))
		}

		guard let viewAnimation = viewAnimation else { return }

		if (viewAnimation.superview == nil) {
			toolbarHUD?.addSubview(viewAnimation)
		}

		viewAnimation.subviews.forEach {
			$0.removeFromSuperview()
		}

		viewAnimation.layer.sublayers?.forEach {
			$0.removeFromSuperlayer()
		}

		if (animationType == .activityIndicator)		{ animationActivityIndicator(viewAnimation)		}
		if (animationType == .ballVerticalBounce)		{ animationBallVerticalBounce(viewAnimation)	}
		if (animationType == .barSweepToggle)			{ animationBarSweepToggle(viewAnimation)		}
		if (animationType == .circleArcDotSpin)			{ animationCircleArcDotSpin(viewAnimation)		}
		if (animationType == .circleBarSpinFade)		{ animationCircleBarSpinFade(viewAnimation)		}
		if (animationType == .circleDotSpinFade)		{ animationCircleDotSpinFade(viewAnimation)		}
		if (animationType == .circlePulseMultiple)		{ animationCirclePulseMultiple(viewAnimation)	}
		if (animationType == .circlePulseSingle)		{ animationCirclePulseSingle(viewAnimation)		}
		if (animationType == .circleRippleMultiple)		{ animationCircleRippleMultiple(viewAnimation)	}
		if (animationType == .circleRippleSingle)		{ animationCircleRippleSingle(viewAnimation)	}
		if (animationType == .circleRotateChase)		{ animationCircleRotateChase(viewAnimation)		}
		if (animationType == .circleStrokeSpin)			{ animationCircleStrokeSpin(viewAnimation)		}
		if (animationType == .dualDotSidestep)			{ animationDualDotSidestep(viewAnimation)		}
		if (animationType == .horizontalBarScaling)		{ animationHorizontalBarScaling(viewAnimation)	}
		if (animationType == .horizontalDotScaling)		{ animationHorizontalDotScaling(viewAnimation)	}
		if (animationType == .pacmanProgress)			{ animationPacmanProgress(viewAnimation)		}
		if (animationType == .quintupleDotDance)		{ animationQuintupleDotDance(viewAnimation)		}
		if (animationType == .semiRingRotation)			{ animationSemiRingRotation(viewAnimation)		}
		if (animationType == .sfSymbolBounce)			{ animationSFSymbolBounce(viewAnimation)		}
		if (animationType == .squareCircuitSnake)		{ animationSquareCircuitSnake(viewAnimation)	}
		if (animationType == .triangleDotShift)			{ animationTriangleDotShift(viewAnimation)		}
	}
}

// MARK: - Setup Sizes
extension ProgressHUD {

	private func setupSizes(_ text: String?, _ animation: Bool) {
		if let text {
			if (animation) && (animationType == .none) {
				setupSizesTextOnly(text)
			} else {
				setupSizesBoth(text)
			}
		} else {
			setupSizesTextNone()
		}
	}

	private func setupSizesBoth(_ text: String) {
		var rect = rectText(text)
		let base = mediaSize + 2 * marginSize

		let width = max(base, rect.size.width + 2 * marginSize)
		let height = max(base, rect.size.height + 2 * marginSize + mediaSize)

		let center = CGPoint(x: width / 2, y: marginSize + mediaSize / 2)

		rect.origin.x = (width - rect.size.width) / 2
		rect.origin.y = (height - rect.size.height) / 2 + (mediaSize + marginSize) / 2

		setupSizes(width, height, center, rect)
	}

	private func setupSizesTextOnly(_ text: String) {
		var rect = rectText(text)
		let base = mediaSize + 2 * marginSize

		let width = max(base, rect.size.width + 2 * marginSize)
		let height = max(base, rect.size.height + 2 * marginSize)

		rect.origin.x = (width - rect.size.width) / 2
		rect.origin.y = (height - rect.size.height) / 2

		setupSizes(width, height, CGPointZero, rect)
	}

	private func setupSizesTextNone() {
		let width = mediaSize + 2 * marginSize
		let height = mediaSize + 2 * marginSize

		let center = CGPoint(x: width / 2, y: height / 2)

		setupSizes(width, height, center, CGRectZero)
	}

	private func setupSizes(_ width: CGFloat, _ height: CGFloat, _ center: CGPoint, _ rect: CGRect) {
		toolbarHUD?.bounds = CGRect(x: 0, y: 0, width: ceil(width), height: ceil(height))

		viewProgress?.center = center
		viewLiveIcon?.center = center
		viewStaticImage?.center = center
		viewAnimation?.center = center

		labelStatus?.frame = rect
	}

	private func rectText(_ text: String) -> CGRect {
		let size = CGSize(width: 250, height: 250)
		let attributes = [NSAttributedString.Key.font: fontStatus]
		return text.boundingRect(with: size, options: .usesLineFragmentOrigin, attributes: attributes, context: nil)
	}
}

// MARK: - Setup Position
extension ProgressHUD {

	@objc private func setupPosition(_ notification: Notification? = nil) {
		var heightKeyboard: CGFloat = 0
		var animationDuration: TimeInterval = 0

		if let notification = notification {
			let frameKeyboard = notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect ?? CGRect.zero
			animationDuration = notification.userInfo?[UIResponder.keyboardAnimationDurationUserInfoKey] as? TimeInterval ?? 0

			if (notification.name == keyboardWillShow) || (notification.name == keyboardDidShow) {
				heightKeyboard = frameKeyboard.size.height
			} else if (notification.name == keyboardWillHide) || (notification.name == keyboardDidHide) {
				heightKeyboard = 0
			} else {
				heightKeyboard = keyboardHeight()
			}
		} else {
			heightKeyboard = keyboardHeight()
		}

		DispatchQueue.main.async { [self] in
			let center = CGPoint(x: main.bounds.size.width / 2, y: (main.bounds.size.height - heightKeyboard) / 2)

			UIView.animate(withDuration: animationDuration, delay: 0, options: .allowUserInteraction) { [self] in
				viewBackground?.frame = main.bounds
				toolbarHUD?.center = center
			}
		}
	}

	private func keyboardHeight() -> CGFloat {
		let windows = UIApplication.shared.windows
		for window in windows {
			for view in window.subviews {
				if String(describing: type(of: view)).hasPrefix("UIInputSetContainerView") {
					for subview in view.subviews {
						if String(describing: type(of: subview)).hasPrefix("UIInputSetHostView") {
							let screenRect = UIScreen.main.bounds
							let keyboardRect = window.convert(subview.frame, to: nil)
							if keyboardRect.intersects(screenRect) {
								return subview.frame.height
							}
						}
					}
				}
			}
		}
		return 0
	}
}

// MARK: - Display, Dismiss, Remove, Destroy
extension ProgressHUD {

	private func displayHUD() {
		if (alpha == 0) {
			alpha = 1
			toolbarHUD?.alpha = 0
			toolbarHUD?.transform = CGAffineTransform(scaleX: 1.4, y: 1.4)

			UIView.animate(withDuration: 0.15, delay: 0, options: [.allowUserInteraction, .curveEaseIn], animations: { [self] in
				toolbarHUD?.transform = CGAffineTransform(scaleX: 1/1.4, y: 1/1.4)
				toolbarHUD?.alpha = 1
			}, completion: nil)
		}
	}

	func dismissHUD() {
		if (alpha == 1) {
			UIView.animate(withDuration: 0.15, delay: 0, options: [.allowUserInteraction, .curveEaseIn], animations: { [self] in
				toolbarHUD?.transform = CGAffineTransform(scaleX: 0.3, y: 0.3)
				toolbarHUD?.alpha = 0
			}, completion: { [self] _ in
				destroyHUD()
				alpha = 0
			})
		}
	}

	func removeHUD() {
		if (alpha == 1) {
			toolbarHUD?.alpha = 0
			destroyHUD()
			alpha = 0
		}
	}

	private func destroyHUD() {
		removeDelayTimer()
		removeNotifications()

		removeLiveIcon()
		removeStaticImage()
		removeProgressView()
		removeAnimationView()

		removeStatus()
		removeToolbar()
		removeBackground()
	}
}
